libtenzir/src/tql2/eval_impl.cpp

```diff
@@ -254,35 +253,44 @@ auto evaluator::eval(const ast::index_expr& x) -> multi_series {
       return eval(ast::field_access{
         x.expr,
         location::unknown,
-        x.suppress_warnings,
+        x.has_question_mark,
         ast::identifier{*string, constant->source},
       });
     }
   }
   return map_series(
     eval(x.expr), eval(x.index),
     [&](series value, const series& index) -> multi_series {
+      const auto add_suppress_hint = [&](auto dh) {
+        if (x.rbracket != location::unknown) {
+          dh = std::move(dh).hint("use `[…]?` to suppress this warning");
+        }
+        return dh;
+      };
       TENZIR_ASSERT(value.length() == index.length());
       if (auto null = value.as<null_type>()) {
-        if (not x.suppress_warnings) {
+        if (not x.has_question_mark) {
           diagnostic::warning("tried to access field of `null`")
-            .primary(x.expr)
+            .primary(x.expr, "is null")
+            .compose(add_suppress_hint)
             .emit(ctx_);
         }
         return std::move(*null);
       }
       if (auto null = index.as<null_type>()) {
-        diagnostic::warning("cannot use `null` as index")
-          .primary(x.index)
-          .emit(ctx_);
+        if (not x.has_question_mark) {
+          diagnostic::warning("cannot use `null` as index")
+            .primary(x.index, "is null")
+            .compose(add_suppress_hint)
+            .emit(ctx_);
+        }
         return std::move(*null);
       }
       if (auto str = index.as<string_type>()) {
         auto* ty = try_as<record_type>(value.type);
         if (not ty) {
           diagnostic::warning("cannot access field of non-record type")
-            .primary(x.index)
-            .secondary(x.expr, "type `{}`", value.type.kind())
+            .primary(x.expr, "has type `{}`", value.type.kind())
```

Comment:
You changed the primary location from `x.index` to `x.expr`. Shouldn't it stay at the index, because the preceding expression itself is fine?

Reply:
I reverted this in 56acb971106.